MobileWallet/Screens/Home/TransactionHistory/TransactionsTableViewControllerExtension.swift

```diff
@@ -70,32 +70,141 @@ extension TransactionsTableViewController {
+        }
+
+        return count
     }
 
     override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
-        return transactions[section].count
+        //If it's the first group and we're showing the pending group
+        if section == 0 && showPendingGroup {
+            return pendingInboundTransactions.count + pendingOutboundTransactions.count
+        }
+
+        let index = showPendingGroup ? section - 1 : section
+
+        return groupedCompletedTransactions[index].count
     }
 
     override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
         let cell = tableView.dequeueReusableCell(withIdentifier: CELL_IDENTIFIER, for: indexPath) as! TransactionTableTableViewCell
 
-        let transaction = transactions[indexPath.section][indexPath.row]
+        var message = ""
+        var contact: Contact?
+        var microTari: MicroTari?
+        var direction: TransactionDirection = .inbound
+
+        //If it's the first group and we're showing the pending group
+        if indexPath.section == 0 && showPendingGroup {
+            if indexPath.row < pendingInboundTransactions.count {
+                let inboundTransaction = pendingInboundTransactions[indexPath.row]
+
+                message = inboundTransaction.message.0
+                contact = inboundTransaction.contact.0
+                microTari = inboundTransaction.microTari.0
+                direction = inboundTransaction.direction
+            } else {
+                let outboundTransaction = pendingOutboundTransactions[indexPath.row - pendingInboundTransactions.count]
+
+                message = outboundTransaction.message.0
+                contact = outboundTransaction.contact.0
+                microTari = outboundTransaction.microTari.0
+                direction = outboundTransaction.direction
+            }
+        } else {
+            //Handle as a completed transaction
+            let index = showPendingGroup ? indexPath.section - 1 : indexPath.section
+
+            let transaction = groupedCompletedTransactions[index][indexPath.row]
+
+            message = transaction.message.0
+            contact = transaction.contact.0
+            microTari = transaction.microTari.0
+            direction = transaction.direction
+        }
+
+        cell.icon.image = Theme.shared.images.transfer
 
-        cell.icon.image = transaction.icon
-        cell.userNameLabel.text = transaction.userName
-        cell.descriptionLabel.text = transaction.description
+        var userName = "-"
+        if let c = contact {
+            let (alias, aliasError) = c.alias
+            userName = alias
+        }
 
-        cell.setValueLabel(tariValue: transaction.value)
+        cell.userNameLabel.text = userName
+        cell.setMessage(message != "" ? message : "*Missing note*")
+        cell.setValueLabel(microTari: microTari != nil ? microTari! : MicroTari(0), direction: direction)
```

Comment:
I usually refactor all my display logic to function(s) in the cell class - TransactionTableTableViewCell in this case. Might be a matter of taste tho.

Reply:
Added 3 different `setDetails` functions to my cell class that each accepts a different transactions type. Thanks that cleans up the table view VC nicely.